File: src/main/java/org/thermocutie/thermostat/device/TemperatureSensor.java
package org.thermocutie.thermostat.device;

import org.thermocutie.thermostat.listener.composite.TemperatureCompositeListener;
import org.thermocutie.thermostat.model.Temperature;
import org.thermocutie.thermostat.mqtt.IMqttListener;
import org.w3c.dom.Element;

/**
 * Temperature sensor device
 *
 * @author Eugene Schava <eschava@gmail.com>
 */
public class TemperatureSensor extends TemperatureCompositeListener
        implements IDevice, IMqttListener {
    private String name;
    private String mqttBroker; // TODO: maybe move all MQTT-related stuff to another class?
    private String mqttTopic; // TODO: maybe move all MQTT-related stuff to another class?

    public TemperatureSensor() {
        super(true);
    }

    @Override
    public String getName() {
        return name;
    }

    public void setMqttTopic(String mqttTopic) {
        this.mqttTopic = mqttTopic;
    }

    @Override
    public String getMqttTopic() {
        return mqttTopic;
    }

    @Override
    public String getMqttBroker() {
        return mqttBroker;
    }

    @Override
    public void onMessage(String payload) {
        Temperature temperature = new Temperature();
        temperature.setValue(Double.parseDouble(payload));

        onTemperatureChanged(temperature);
    }

    @Override
    public void loadFromXml(Element element) {
        name = getAttribute(element,"name");
        mqttBroker = getAttribute(element,"mqttBroker");
        mqttTopic = getAttribute(element,"mqttTopic");
    }

    @Override
    public void saveToXml(Element element) {
        element.setAttribute("name", name);
        element.setAttribute("mqttBroker", mqttBroker);
        element.setAttribute("mqttTopic", mqttTopic);
    }
}

File: python/domain/device/TemperatureSensor.py
from .Device import Device


class TemperatureSensor(Device):
    TYPE = 'Temperature'

    def __init__(self):
        Device.__init__(self, self.TYPE)
        self._mqttBroker = None
        self._mqttTopic = None

    @property
    def mqtt_broker(self):
        return self._mqttBroker

    @property
    def mqtt_topic(self):
        return self._mqttTopic

    def load(self, xml):
        super(TemperatureSensor, self).load(xml)

        self._mqttBroker = xml.attrib.get('mqttBroker', None)
        self._mqttTopic = xml.attrib.get('mqttTopic', None)

    def save(self, xml):
        super(TemperatureSensor, self).save(xml)

        if self._mqttBroker:
            xml.attrib['mqttBroker'] = self._mqttBroker
            xml.attrib['mqttTopic'] = self._mqttTopic
        else:
            del xml.attrib['mqttBroker']
            del xml.attrib['mqttTopic']

    def json(self):
        json = super(TemperatureSensor, self).json()
        json.update({'mqtt_broker': self._mqttBroker, 'mqtt_topic': self._mqttTopic})
        return json

    def update(self, changes):
        self._name = changes['name']
        if 'mqtt_broker' in changes:
            self._mqttBroker = changes['mqtt_broker']
            self._mqttTopic = changes['mqtt_topic']
        else:
            self._mqttBroker = None
            self._mqttTopic = None

